### PR TITLE
[WIP] Feature image data

### PR DESCRIPTION
**Pull request by kalwalt**
# Feature image data
This PR continue from issue #54 and try to find a solution to export `ImageData` from a `sink` node.
In the issue i tested to achieve this task in an extern class outside speedy-vision but this was failing. Instead i extended the SpeedyMedia type to accept ImageData and i created a new class **SpeedyImageDataMediaSource** that extends **SpeedyMediaSource** class and then in the Image.Sink node i created a new exportImageData function to export ImageData. I tested it and it export an ImageData object, but there are some considerations to do:

1. maybe we can export the ImageData in the export method instead.
2. as a consequence of the first point add plus options to export() ( export ImageBitmap, ImageData or other supported `Speedy MediaType's` if this make sense)
3. add some tests

For this reason i send a draft Pull request.

### CODE REVIEW

File: src/core/speedy-media-source.js [hunk 1]
```diff
@@ -702,11 +704,119 @@ class SpeedyBitmapMediaSource extends SpeedyMediaSource
 
     /**
      * Load the underlying media
-     * @param {ImageBitmap} bitmap
+     * @param {ImageData} bitmap
```
Comment (kalwalt) [inline src/core/speedy-media-source.js:707]: I need to revert this to ImageBitmap sorry for this!

File: src/core/speedy-media-source.js [hunk 2]
```diff
@@ -702,11 +704,119 @@ class SpeedyBitmapMediaSource extends SpeedyMediaSource
...
+
+    /**
+     * Clone this media source
+     * @returns {SpeedyPromise<SpeedyMediaSource>}
+     */
+    clone()
+    {
+        if(this._data == null)
+            throw new IllegalOperationError(`Media not loaded`);
+
```
Comment (kalwalt) [inline src/core/speedy-media-source.js:774]: i'm note sure that this code actually is cloning or copying, i need to check.

File: types/core/speedy-media-source.d.ts [hunk 3]
```diff
@@ -73,5 +73,5 @@ export class SpeedyMediaSource {
      */
     _load(element: SpeedyMediaSourceNativeElement): SpeedyPromise<SpeedyMediaSource>;
 }
-export type SpeedyMediaSourceNativeElement = HTMLImageElement | HTMLVideoElement | HTMLCanvasElement | ImageBitmap;
+export type SpeedyMediaSourceNativeElement = HTMLImageElement | HTMLVideoElement | HTMLCanvasElement | OffscreenCanvas | ImageBitmap;
```
Comment (kalwalt) [inline types/core/speedy-media-source.d.ts:76]: i think i should add also ImageData to SpeedyMediaSourceNativeElement. I noticed that tsc added some changes, maybe tpes definitions wasn't updated?

**Comment by alemart** [comment]
Adding support for an `ImageData` as a `SpeedyMedia` is a lot simpler than that. It's easy to do: take a look at PR #51.

Once that's done and merged, it will also be easy to export a `SpeedyMedia` holding an `ImageData` with the `readPixelsAsync` method I explained you in #54.

No need to change all these files.

**Comment by kalwalt** [comment]
> Adding support for an `ImageData` as a `SpeedyMedia` is a lot simpler than that. It's easy to do: take a look at PR #51.
> 

I think i have done It with the first commit. So you are saying it' s not necessary the exportImageData() in the sink node.

> Once that's done and merged, it will also be easy to export a `SpeedyMedia` holding an `ImageData` with the `readPixelsAsync` method I explained you in #54.
>

There Is some example somewhere?


> No need to change all these files.

Yes, i understood! 🙂

**Comment by kalwalt** [comment]
I think i will close this PR and reopen a new one with only changes for adding support for an `ImageData` as a `SpeedyMedia`, i'm not sure how to deal with the readPixelsAsync and maybe we can develop it in further steps with your help.

**Comment by kalwalt** [comment]
Closing, new updated PR is #58

**Pull request closed without merging**